File: pdf-rag-ts/src/prompting/query_augmentation.ts
export class QueryAugmenter {
    private ollamaUrl: string;
    private generativeModel: string;
    private rewriteModel: string;

    constructor(ollamaUrl: string, generativeModel: string, rewriteModel: string) {
        this.ollamaUrl = ollamaUrl;
        this.generativeModel = generativeModel;
        this.rewriteModel = rewriteModel;
    }

    async rewriteQuery(question: string): Promise<string> {
        const prompt = `You are a search query optimizer. Your task is to provide 3-5 high-quality keywords or technical synonyms that would help find the answer to the user's question in a technical document.

Rules:
- Output ONLY the keywords, separated by spaces.
- Do NOT repeat words already in the question.
- Do NOT include conversational filler.
- Focus on technical terminology.

Question: what is RAG
Keywords: retrieval augmented generation vector database embeddings

Question: how does chunking work
Keywords: splitting segmentation tokenization overlap context

Question: what are the requirements
Keywords: prerequisites specifications criteria qualifications

Question: ${question}
Keywords:`;

        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), 30000);

        try {
            const response = await fetch(`${this.ollamaUrl}/api/generate`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    model: this.rewriteModel,
                    prompt: prompt,
                    stream: false,
                }),
                signal: controller.signal,
            });

            clearTimeout(timeoutId);

            if (!response.ok) {
                return question;
            }

            const data = await response.json() as { response?: string };
            const expansion = (data.response || "").trim().toLowerCase();
            
            // Filter out any common conversational words the model might have included
            const filteredExpansion = expansion.replace(/^(keywords:|expanded:)/i, '').trim();
            
            const rewritten = filteredExpansion ? `${question} ${filteredExpansion}` : question;
            console.log(`🔄 Query expanded: '${question}' → '${rewritten}'`);
            return rewritten;
        } catch (error) {
            clearTimeout(timeoutId);
            return question;
        }
    }

    async generateHypotheticalAnswer(question: string): Promise<string> {
        const prompt = `Write a short, technically detailed passage (2-4 sentences) that directly answers the following question. Write as if you are the document being searched — use specific terms, model names, and technical details that would appear in a technical document.

Question: ${question}
Passage:`;

        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), 30000);

        try {
            const response = await fetch(`${this.ollamaUrl}/api/generate`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    model: this.generativeModel,
                    prompt: prompt,
                    stream: false,
                }),
                signal: controller.signal,
            });

            clearTimeout(timeoutId);

            if (!response.ok) {
                return question;
            }

            const data = await response.json() as { response?: string };
            const hypothetical = (data.response || question).trim();
            console.log(`💭 HyDE passage: ${hypothetical.substring(0, 100)}...`);
            return hypothetical;
        } catch (error) {
            clearTimeout(timeoutId);
            return question;
        }
    }
}

File: pdf-rag-python/src/prompting/query_augmentation.py
import requests


class QueryAugmenter:
    def __init__(self, ollama_url: str, generative_model: str, rewrite_model: str):
        self.ollama_url = ollama_url
        self.generative_model = generative_model
        self.rewrite_model = rewrite_model

    def rewrite_query(self, question: str) -> str:
        prompt = f"""You are a search query expander for document retrieval. Your job is to add synonyms and closely related terms to help find the answer in any type of document.

Rules:
- Only add terms that are directly related to what the question is asking about.
- For short factual questions ("What is X?", "Who is Y?"), keep the expansion tight — 3 to 5 terms maximum.
- Never invent terms that could belong to a different topic entirely.
- Return only the expanded query. Use English only.

Question: what is RAG
Expanded: RAG retrieval augmented generation pipeline architecture

Question: how does chunking work
Expanded: chunking text splitting document segmentation sentence splitting chunk size overlap

Question: what is the main topic
Expanded: main topic subject purpose overview summary

Question: what are the requirements
Expanded: requirements qualifications criteria conditions prerequisites

Question: how to reduce hallucination
Expanded: hallucination reduction grounding faithfulness factuality

Question: {question}
Expanded:"""
        try:
            response = requests.post(
                f"{self.ollama_url}/api/generate",
                json={"model": self.rewrite_model, "prompt": prompt, "stream": False},
                timeout=30,
            )
            response.raise_for_status()
            expansion = response.json().get("response", "").strip()
            # Always keep original terms — expansion is additive, never a replacement.
            # This ensures named features/proper nouns from the question survive even
            # when the small rewrite model paraphrases them away.
            rewritten = f"{question} {expansion}" if expansion else question
            print(f"🔄 Query expanded: '{question}' → '{rewritten}'")
            return rewritten
        except Exception:
            return question

    def generate_hypothetical_answer(self, question: str) -> str:
        prompt = f"""Write a short, technically detailed passage (2-4 sentences) that directly answers the following question. Write as if you are the document being searched — use specific terms, model names, and technical details that would appear in a technical document.

Question: {question}
Passage:"""
        try:
            response = requests.post(
                f"{self.ollama_url}/api/generate",
                json={"model": self.generative_model, "prompt": prompt, "stream": False},
                timeout=30,
            )
            response.raise_for_status()
            hypothetical = response.json().get("response", question).strip()
            print(f"💭 HyDE passage: {hypothetical[:100]}...")
            return hypothetical
        except Exception:
            return question
